File: ios/Khrysalis/com/lightningkite/khrysalis/Box.actual.swift
import Foundation

public protocol BoxProtocol {
    associatedtype T
    func asBox() -> Box<T>
}

//--- Box.{
public struct Box<T>: BoxProtocol {

    public func asBox() -> Box<T> {
        return self
    }

    //--- Box.Primary Constructor
    public init(_ value: T) {
        self.value = value
    }

    //--- Box.value
    public var value: T

    //--- Box.Companion.{

    //--- Box.Companion.wrap(T)
    public static func wrap(_ value: T) -> Box<T> {
        return Box(value)
    }
    public static func wrap(value: T) -> Box<T> {
        return Box(value)
    }

    //--- Box.Companion.}

    //--- Box.}
}

//--- boxWrap(T)
public func boxWrap<T>(_ value: T) -> Box<T> {
    return Box(value)
}
public func boxWrap<T>(value: T) -> Box<T> {
    return Box(value)
}

File: android/src/main/java/com/lightningkite/khrysalis/Box.actual.kt
package com.lightningkite.khrysalis

/*inline*/ class Box<out T> @Deprecated("Never, ever use this.  Use wrap instead.") constructor(val valueRaw: Any?) {
    @Suppress("UNCHECKED_CAST")
    val value: T get() = if(valueRaw == NullRep) null as T else valueRaw as T
    companion object {
        private val NullRep = object {}
        private val NullRepBox = Box<Any?>(NullRep)
        @Suppress("UNCHECKED_CAST")
        fun <T> wrap(value: T): Box<T> = if(value == null) NullRepBox as Box<T> else Box(value)
    }

    override fun toString(): String {
        return "Box($value)"
    }
}

fun <T> boxWrap(value: T): Box<T> {
    return Box.wrap(value)
}
